modules/tftp/server.rb

```diff
@@ -150,6 +150,36 @@ def pxeconfig_file(mac)
     end
   end
 
+  def self.fetch_boot_image(image_dst, url, files)
+    # Verify image_dst does not contain ".." (switch folder)
+    raise "TFTP image file extraction error: #{file_path}" unless image_dst.split("..").size <= 1
+
+    base_path = Pathname.new("/var/www/html/pub/installation_media")
```

Comment:
Hardcoded paths are always bad. This must at the very least be a setting. But there's more. `/var/www/html` is a Red Hat-ism and `/var/www/html/pub` is a Katello-ism. In fact, there being some HTTP server at all is something that's not guaranteed at all since only Katello does that, but vanilla Foreman doesn't.

I'd say that we should not do this and the Smart Proxy itself should serve the files. Otherwise the feature isn't available on non-Katello and this is a core Smart Proxy module.

Having said that, today we don't have a setting for Foreman Proxy's state directory (essential `/var/lib/foreman-proxy` in production). We've ran into that a few times and would be a good feature. Perhaps this is the excuse to start introducing that.

Reply:
Alright, thanks for this comment. It makes totally sense to have this as a core Smart Proxy feature to guarantee that it works independent of the environment (Katello installed or not).
Can you maybe elaborate a bit more on how the Smart Proxy itself can serve the files? And how is this connected to the state directory? I assume you don't want to run a web server offering the files in `/var/lib/foreman-proxy/`.

Moreover, if the directory where the files are hosted is a dynamic setting on the Smart Proxy, the corresponding address might be dynamic as well (not just `http://my.proxy/installation_media/..`, but depending on the Smart Proxy configuration). How can we sync this information to the Foreman side of the plugin? This is necessary because the boot template needs the full URL of the ISO image. 
So, is there any tool to sync Smart Proxy settings with the Foreman and make them accessible during runtime?

Thanks for all ur help on this btw. It's highly appreciated! :raised_hands:

Reply:
> Can you maybe elaborate a bit more on how the Smart Proxy itself can serve the files? And how is this connected to the state directory? I assume you don't want to run a web server offering the files in /var/lib/foreman-proxy/.

The Smart Proxy **is** a webserver. That's how it also serves httpboot files. See https://github.com/theforeman/smart-proxy/blob/develop/modules/httpboot/httpboot_api.rb for an example.

> Moreover, if the directory where the files are hosted is a dynamic setting on the Smart Proxy, the corresponding address might be dynamic as well (not just http://my.proxy/installation_media/.., but depending on the Smart Proxy configuration). How can we sync this information to the Foreman side of the plugin? This is necessary because the boot template needs the full URL of the ISO image.

I'd suggest that the HTTP call to download them has the resulting paths in the HTTP response. There's no way to know the filenames from an archive.

An interesting thought is idempotency and concurrency. What if the system tries to download the same ISO multiple times. Does it download it again or reuses? If it keeps files, how does it know if it's still up to date? This is also a problem we have with the downloaded kernel files so it may be fine to ignore for now and assume ISOs never change. Just want you to be aware of it.

>  So, is there any tool to sync Smart Proxy settings with the Foreman and make them accessible during runtime?

Settings can be exposed. Again, https://github.com/theforeman/smart-proxy/blob/develop/modules/httpboot/httpboot_plugin.rb is an interesting example.

Reply:
Thanks for all these examples! I've mostly taken the additional API endpoint as it's implemented in `httpboot` and added the endpoint `/tftp/boot_image/`. You can see the "most important" diff [here](https://github.com/theforeman/smart-proxy/compare/2fa76995ae193158e6d59d3659661ecadb5c1c70..a0ee76d1fbf7fc7f2e2c9d6b8648e9f1e2dd3e9f).

I think it is important to mention that the plugin setting in `/etc/foreman-proxy/settings.d/tftp.yml` must be:
```
:enabled: true
...
```
Instead of just `https` since both protocols are used from now on.

I think idempotency and concurrency is an important topic here, since the ISO image download takes a while which increases the probability that another download process is triggered while the first one is not finished. As of now, the second process is just going to fail if it identifies that the ISO is currently being downloaded/extracted (have a look [here](https://github.com/theforeman/smart-proxy/pull/837/files#diff-db3f0d8d32c77f2719de93fd56902295e1d21a1a1c0dc1e19f398cf94d69426eR167)).
In terms of ISO image "version" (do we have the latest ISO downloaded?), there is currently no specific strategy. If an image has been downloaded, this one going to be used. A re-download can be "triggered" by deleting the image manually.